src/HtmlReader/architecture.md

```diff
@@ -0,0 +1,78 @@
+
+Readium-CSS also enables a "paginated" mode, where the full HTML resource is broken into multiple columns that are then rendered side by side in a horizontally scrolling container. Thus, to move to the next page in paginated mode, we simply scroll the user horizontally by one page-width.
+
+## Lifecycle and State
+
+We use `React.useReducer` as the single source of truth for state in this hook. Whenever anything changes, it should go through `useReducer`. If you render with the same state object twice, you should always get the same thing.
+
+Given that there are more than 10 properties in the state object, and each one has multiple possible values, there are seemingly infinite possible states. However, the range of _valid_ states is actually much smaller. These states are:
+
+- `InactiveState`: The hook was not provided any arguments, and thus it is inactive.
+- `FetchingResourceState`: We are fetching a resource, and thus are "loading".
+- `ResourceFetchErrorState`: We have encountered an error while fetching the resource.
+- `RenderingIframeState`: We have the resource, and are now waiting for the iframe to finish rendering so that we have a reference to it. We get this reference via the `ref` property on the `<iframe>` tag. This is a very short state as it only exists during the single render cycle between when the resource finishes fetching and the iframe renders for the first time. Note that rendering the iframe element does not mean the iframe has rendered its children, that is separate.
+- `LoadingIframeState`: We have rendered the iframe once, and are now waiting for it to fire off it's own `window.onLoad` function indicating that it has loaded all the html in the resource string we passed it.
+- `NavigatingState`: The iframe has finished loading, and now we are waiting for our `useEffect` hook to complete any inter-resource navigation that might be necessary depending on the current `state.location`. For example, scrolling the user to a specific page.
+- `ReadyState`: We have now navigated the user to a specific page, and are now finished with our work.
+
+The definitions for these states can be found in `./types.ts`. Note that state transitions should happen in the order listed above, with the exception that it is possible to go backwards occasionally. For example, from the `ReadyState`, we may go back to `FetchingResourceState` if the user switches to a new resource, or we might simply go back to the `NavigatingState` if they navigate to a new location in the same resource.
+
+We provide the text value `state.state` as a string describing the state. This makes it very easy to perform type narrowing. If `state.state === 'READY'`, typescript knows that `state.iframe` and `state.resource`, and more all exist.
```

Comment:
should those values stay strings or should we enum them

Reply:
Enums aren't very nice in TS unfortunately, and strings provide all the same benefits. They are _typed_ strings though, so you can't accidentally put something different in there